scripts/generate_hamilton_input_qpcr_dilution.py

```diff
@@ -0,0 +1,112 @@
+                output.udf['RSB Volume (ul)'] = str(int(round(total_volume - float(self.step_udfs['DCT Volume (ul)']),1)))
+            except:
+                raise InvalidStepError('Missing value. Check that step UDFs (Adjusted Conc. (nM) and Target '
+                                       'Concentration (nM)) are populated')
+        else:
+
+            output.udf['RSB Volume (ul)'] = '0'
+        self.artifacts_to_update.add(output)
+        return output
+
+    def _generate_csv_dict(self):
+        # csv_dict will be a dictionary that consists of the lines to be present in the Hamilton input file.
+        csv_dict = {}
+
+        # find the lot number, i.e. barcode, of the RSB reagent.
+        buffer_template = "LP[0-9]{7}-RSB"
+        reagent_lots = list(self.process.step.reagent_lots)
+
+        buffer_barcode = None
+        for lot in reagent_lots:
+            if re.match(buffer_template, lot.lot_number):
+                buffer_barcode = lot.lot_number
+
+        if not buffer_barcode:
+            raise InvalidStepError('Please assign buffer lot before generating Hamilton input.')
+
+        # find all the inputs for the step that are analytes (i.e. samples and not associated files)
+        for input_art in self.artifacts:
+            if input_art.type == 'Analyte':
+                output = self.calculate_rsb_volume(input_art)
+
+                # remove semi-colon from locations as this is not compatible with Hamilton Venus software
+                input_location = input_art.location[1].replace(':', '')
+
+                # assemble each line of the Hamilton input file in the correct structure for the Hamilton
+                # in this step the barcode is transferred from the input to the output plate so no new analyte is
+                # created and not output location exists. The transfer is 1:1 so therefore the input container and input
+                #  location are used for both input and output locations in the hamilton input file
+                csv_line = [
+                    input_art.container.name, input_location, input_art.container.name, input_location,
+                    self.step_udfs['DCT Volume (ul)'], buffer_barcode, output.udf['RSB Volume (ul)']
+                ]
+                # build a dictionary of the lines for the Hamilton input file with a key that facilitates
+                # the lines being by input container then column then row
+                csv_dict[input_art.location[1]] = csv_line
+
+        return csv_dict
+
+    def _run(self):
+        # create the set to hold the inputs to be updated with the output from the calculation
+        self.artifacts_to_update = set()
+        #input udfs to be transposed to output udfs
+        self.udf_names = ['Adjusted Conc. (nM)', 'Ave. Conc. (nM)', 'Original Conc. (nM)', '%CV']
+
+        self.step_udfs = self.process.udf
```

Comment:
This statement seems unnecessary. `self.step_udfs` is not used in this function

Reply:
It's used by the chain of instance methods "_generate_csv_dict()" and "calculate_rsb_volume()" so it seemed logical to define it here as it can't be defined as a class variable? Is there a better way of doing this?

Reply:
...but because I am using "super" for run, I now need to move it to the first method in the chain..updated

Reply:
statement removed and replaced with self.process.udf